src/System.Private.ServiceModel/tests/Stress/SharedPoolsOfWCFObjects/SharedPoolsOfWCFObjects.cs

```diff
@@ -0,0 +1,612 @@
+                        createObject: (chf) =>
+                            s_CreateChannelStats.CallFuncAndRecordStats(TestHelpers.CreateChannel, chf),
+                        destroyObject: (ch) =>
+                            s_CloseChannelStats.CallActionAndRecordStats(() => TestHelpers.CloseChannel(ch))
+                        ),
+                    destroyInstance: (_fapoic) => _fapoic.Destroy()));
+        }
+
+        public static void RunAllScenariosWithWeights(int useWeight, int recycleChannelsWeight, int recycleFactoriesWeight)
+        {
+            int seed = Interlocked.Increment(ref s_iteration) % (useWeight + recycleChannelsWeight + recycleFactoriesWeight);
+            if (seed < useWeight)
+            {
+                UsePooledChannels();
+            }
+            else if (seed < useWeight + recycleChannelsWeight)
+            {
+                RecyclePooledChannels();
+            }
+            else if (seed < useWeight + recycleChannelsWeight + recycleFactoriesWeight)
+            {
+                RecyclePooledFactories();
+            }
+        }
+
+        public static void UsePooledChannels()
+        {
+            foreach (var factoryAndPoolOfItsChannels in s_recyclablePooledFactoriesAndChannels.GetAllPooledInstances())
+            {
+                foreach (var channel in factoryAndPoolOfItsChannels.ObjectsPool.GetAllPooledInstances())
+                {
+                    s_CallChannelStats.CallActionAndRecordStats(() =>
+                    {
+                        channel.GetData(44);
+                    });
+                }
+            }
+        }
+
+        public static void RecyclePooledChannels()
+        {
+            Console.WriteLine("RecyclePooledChannels");
+            foreach (var factoryAndPoolOfItsChannels in s_recyclablePooledFactoriesAndChannels.GetAllPooledInstances())
+            {
+                factoryAndPoolOfItsChannels.ObjectsPool.DestoryAllPooledInstances();
+            }
+        }
+
+        public static void RecyclePooledFactories()
+        {
+            Console.WriteLine("RecyclePooledFactories");
+            s_recyclablePooledFactoriesAndChannels.DestoryAllPooledInstances();
+            // additional checks - move to DestoryAllPooledInstances itself
+            foreach (var factoryAndPoolOfItsChannels in s_recyclablePooledFactoriesAndChannels.GetAllPooledInstances())
+            {
+                if (factoryAndPoolOfItsChannels.Factory.State == CommunicationState.Closed)
+                {
+                    System.Diagnostics.Debugger.Break();
+                }
+            }
+        }
+    }
+
+    public class RecyclablePooledFactoriesAndChannels_OpenOnce
+    {
+        private static int s_iteration = 0;
+        private static CallStats s_CreateChannelStats = new CallStats(samples: 1000, errorSamples: 1000);
+        private static CallStats s_CallChannelStats = new CallStats(samples: 1000000, errorSamples: 10000);
```

Comment:
For future consideration -- make these magic numbers visible as const's or static readonly fields